### ISSUE TITLE
Use a trained model in a C++ applications?

### ISSUE THREAD

**Issue by pkuppens** [MIC-DKFZ/nnUNet#849]
Does anyone have experience with embedding a trained model in a C++ application?

For instance, is the trained model stored in a way that it can be read/converted to a Tensorflow model, such that we can have Tensorflow do its work?

Which steps would remain? E.g. some input conversion/resizing/interpolation that the preprocessor usually does?

PS. I have not put much effort into investigating this, some pointers would be welcome.

**Comment by wxr521314** [comment]
I have the same problem. Do you solve it?

**Comment by pkuppens** [comment]
@wxr521314 I did not put any effort in this yet. I was hoping for a quick answer of @FabianIsensee (reducing a lot of investigation time on my side), saying something like: file xyz. can be read in Tensorflow/Keras in Python, and should be readable in C++ as well,  similar to https://www.tensorflow.org/tutorials/keras/save_and_load
You still need to provide input in C++ that meets the expectations (image size and format).

I still expect that this will be a big implementation task, but I expect it will be possible.

**Comment by FabianIsensee** [comment]
Hi, this is rather complex to pull off with many potential causes for problems. Here is a rundown:

**obtaining the model**
probably best done via onnx. You can find the parameters in the checkpoint files. You will need the arguments used for initializing the Generic_UNet class. Then you can export to onnx with a dummy input.

**inferencing pipeline - things you need to reimplement**
- cropping to nonzero region
- intensity normalization
- resampling (careful with that!)
- overlapping window prediction with gaussian importance weighting. Important: you cannot do fully convolutional inference! You need to use the same patch size as was used for training
- (optional) ensembling, test time augmentation

Maybe it would be best to simply call the python implementation from within your C++ program?

Best,
Fabian

**Comment by pkuppens** [comment]
Well, @FabianIsensee, if you think it is complex, I don't dare to start this.

Calling the implementation, or having a microservice in place to do so are good workarounds.

Thanks, and I'll close this issue.
